<img src="https://user-images.strikinglycdn.com/res/hrscywv4p/image/upload/c_limit,fl_lossy,h_300,w_300,f_auto,q_auto/1266110/Logo_wzxi0f.png" style="float: left; margin: 20px; height: 55px">

**Tell me what is more important, your life or the adventure? - [Monkey D. Luffy](https://en.wikipedia.org/wiki/Monkey_D._Luffy)**

# 1. Numbers and more in Python

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers



Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Throughout this course we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

 
 
Now let's start with some basic arithmetic.

## Basic Arithmetic

In [1]:
# Addition
2+1 

3

In [2]:
# Subtraction
2-1

1

In [3]:
# Multiplication
2*2

4

In [5]:
# Division
3/2

1.5

In [6]:
# Floor Division
7//4

1

**Whoa! What just happened? Last time I checked, 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

**So what if we just want the remainder after division?**

In [7]:
# Modulo7/4

In [8]:
7%4

3

4 goes into 7 once, with a remainder of 3. The % operator returns the remainder after division.

## Arithmetic continued

In [10]:
# Powers
2**3

8

In [ ]:
# Can also do roots this way
4**0.5

2.0

In [11]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [ ]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

## Variable Assignments

Now that we've seen how to use numbers in Python as a calculator let's see how we can assign names and create variables.

We use a single equals sign to assign labels to variables. Let's see a few examples of how we can do this.

In [12]:
float(2)/3

0.6666666666666666

In [15]:
# Let's create an object called "a" and assign it the number 5
a = 5

Now if I call *a* in my Python script, Python will treat it as the number 5.

In [17]:
# Adding the objects
a + a

10

What happens on reassignment? Will Python let us write it over?

In [18]:
# Reassignment
a = 10

In [19]:
# Check
a

10

Yes! Python allows you to write over assigned variable names. We can also use the variables themselves when doing the reassignment. Here is an example of what I mean:

In [20]:
# Check
a

10

In [ ]:
# Use A to redefine A
a = a + a

In [22]:
# Check 
a

10

The names you use when creating these labels need to follow a few rules:

    1. Names can not start with a number.
    2. There can be no spaces in the name, use _ instead.
    3. Can't use any of these symbols :'",<>/?|\()!@#$%^&*~-+
    4. It's considered best practice (PEP8) that names are lowercase.
    5. Avoid using the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), 
       or 'I' (uppercase letter eye) as single character variable names.
    6. Avoid using words that have special meaning in Python like "list" and "str"


Using variable names can be a very useful way to keep track of different variables in Python. For example:

In [23]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate


In [24]:
# Show my taxes!
my_taxes

10.0

So what have we learned? We learned some of the basics of numbers in Python. We also learned how to do arithmetic and use Python as a basic calculator. We then wrapped it up with learning about Variable Assignment in Python.

Up next we'll learn about Strings!

---

# 2. Variable Assignment

## Rules for variable names

* names can not start with a number
* names can not contain spaces, use _ intead
* names can not contain any of these symbols:

      :'",<>/?|\!@#%^&*~-+
       
* it's considered best practice ([PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are lowercase with underscores
* avoid using Python built-in keywords like `list` and `str`
* avoid using the single characters `l` (lowercase letter el), `O` (uppercase letter oh) and `I` (uppercase letter eye) as they can be confused with `1` and `0`

## Dynamic Typing

Python uses *dynamic typing*, meaning you can reassign variables to different data types. This makes Python very flexible in assigning data types; it differs from other languages that are *statically typed*.

In [25]:
my_dogs = 2

In [26]:
my_dogs

2

In [29]:
my_dogs = ['Sammy', 'Frankie']

In [30]:
my_dogs

['Sammy', 'Frankie']

### Pros and Cons of Dynamic Typing
#### Pros of Dynamic Typing
* very easy to work with
* faster development time

#### Cons of Dynamic Typing
* may result in unexpected bugs!
* you need to be aware of `type()`

## Assigning Variables

Variable assignment follows `name = object`, where a single equals sign `=` is an *assignment operator*

In [ ]:
a = 5

In [ ]:
a

5

Here we assigned the integer object `5` to the variable name `a`.<br>Let's assign `a` to something else:

In [ ]:
a = 10

In [ ]:
a

10

You can now use `a` in place of the number `10`:

In [ ]:
a + a

20

## Reassigning Variables

Python lets you reassign variables with a reference to the same object.

In [31]:
a = a + 10

In [32]:
a

20

There's actually a shortcut for this. Python lets you add, subtract, multiply and divide numbers with reassignment using `+=`, `-=`, `*=`, and `/=`.

In [35]:
a += 10

In [36]:
a

40

In [37]:
a *= 2

In [38]:
a

80

## Determining variable type with `type()`

You can check what type of object is assigned to a variable using Python's built-in `type()` function. Common data types include:
* **int** (for integer)
* **float**
* **str** (for string)
* **list**
* **tuple**
* **dict** (for dictionary)
* **set**
* **bool** (for Boolean True/False)

In [39]:
type(a)

int

In [40]:
a = (1,2)

In [41]:
type(a)

tuple

## Simple Exercise

This shows how variables make calculations more readable and easier to follow.

In [42]:
my_income = 100
tax_rate = 0.1
my_taxes = my_income * tax_rate

In [43]:
my_taxes

10.0

Great! You should now understand the basics of variable assignment and reassignment in Python.<br>Up next, we'll learn about strings!

---

# 3. Strings

Strings are used in Python to record text information, such as names. Strings in Python are actually a *sequence*, which basically means Python keeps track of every element in the string as a sequence. For example, Python understands the string "hello' to be a sequence of letters in a specific order. This means we will be able to use indexing to grab particular letters (like the first letter, or the last letter).

This idea of a sequence is an important one in Python and we will touch upon it later on in the future.

In this lecture we'll learn about the following:

    1.) Creating Strings
    2.) Printing Strings
    3.) String Indexing and Slicing
    4.) String Properties
    5.) String Methods
    6.) Print Formatting

## Creating a String

To create a string in Python you need to use either single quotes or double quotes. For example:

In [48]:
# Single word
'hello'

'hello'

In [49]:
# Entire phrase 
'This is also a string'

'This is also a string'

In [50]:
# We can also use double quote
"String built with double quotes"

'String built with double quotes'

In [54]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-54-da9a34b3dc31>, line 2)

The reason for the error above is because the single quote in <code>I'm</code> stopped the string. You can use combinations of double and single quotes to get the complete statement.

In [52]:
"Now I'm ready to use the single quotes inside a string!"

"Now I'm ready to use the single quotes inside a string!"

Now let's learn about printing strings!

## Printing a String

Using Jupyter notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a print function.

In [55]:
# We can simply declare a string
'Hello World'

'Hello World'

In [56]:
# Note that we can't output multiple strings this way
'Hello World 1'
'Hello World 2'

'Hello World 2'

We can use a print statement to print a string.

In [57]:
print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line')
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?


## String Basics

We can also use a function called len() to check the length of a string!

In [59]:
len('Hello World')

11

Python's built-in len() function counts all of the characters in the string, including spaces and punctuation.

## String Indexing

We know strings are a sequence, which means Python can use indexes to call parts of the sequence. Let's learn how this works.

In Python, we use brackets <code>[]</code> after an object to call its index. We should also note that indexing starts at 0 for Python. Let's create a new object called <code>s</code> and then walk through a few examples of indexing.

In [60]:
# Assign s as a string
s = 'Hello World'

In [61]:
#Check
s

'Hello World'

In [62]:
# Print the object
print(s) 

Hello World


Let's start indexing!

In [63]:
# Show first element (in this case a letter)
s[0]

'H'

In [64]:
s[1]

'e'

In [65]:
s[2]

'l'

We can use a <code>:</code> to perform *slicing* which grabs everything up to a designated point. For example:

In [66]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[1:]

'ello World'

In [67]:
# Note that there is no change to the original s
s

'Hello World'

In [68]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

Note the above slicing. Here we're telling Python to grab everything from 0 up to 3. It doesn't include the 3rd index. You'll notice this a lot in Python, where statements and are usually in the context of "up to, but not including".

In [69]:
#Everything
s[:]

'Hello World'

We can also use negative indexing to go backwards.

In [70]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [71]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [72]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [73]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

In [74]:
# We can use this to print a string backwards
s[::-1]

'dlroW olleH'

## String Properties

It's important to note that strings have an important property known as *immutability*. This means that once a string is created, the elements within it can not be changed or replaced. For example:

In [75]:
s

'Hello World'

In [76]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Notice how the error tells us directly what we can't do, change the item assignment!

Something we *can* do is concatenate strings!

In [79]:
s

'Hello World'

In [80]:
# Concatenate strings!
s + ' concatenate me!'

'Hello World concatenate me!'

In [81]:
# We can reassign s completely though!
s = s + ' concatenate me!'

In [82]:
print(s)

Hello World concatenate me!


In [83]:
s

'Hello World concatenate me!'

We can use the multiplication symbol to create repetition!

In [84]:
letter = 'z'

In [85]:
letter*10

'zzzzzzzzzz'

## Basic Built-in String methods

Objects in Python usually have built-in methods. These methods are functions inside the object (we will learn about these in much more depth later) that can perform actions or commands on the object itself.

We call methods with a period and then the method name. Methods are in the form:

object.method(parameters)

Where parameters are extra arguments we can pass into the method. Don't worry if the details don't make 100% sense right now. Later on we will be creating our own objects and functions!

Here are some examples of built-in methods in strings:

In [86]:
s

'Hello World concatenate me!'

In [87]:
# Upper Case a string
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [88]:
# Lower case
s.lower()

'hello world concatenate me!'

In [89]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [90]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld concatenate me!']

There are many more methods than the ones covered here. Visit the Advanced String section to find out more!

## Print Formatting

We can use the .format() method to add formatted objects to printed string statements. 

The easiest way to show this is through an example:

In [101]:
'Insert another string with curly brackets: {0}, {1}'.format('The inserted string', 'perro')

'Insert another string with curly brackets: The inserted string, perro'

We will revisit this string formatting topic in later sections when we are building our projects!

---

# 4. String Formatting

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them here.

## Formatting with placeholders

You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [103]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [104]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [105]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### Format conversion methods.

It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [106]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [109]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [111]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### Padding and Precision of Floating Point Numbers

Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [122]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [120]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [116]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [117]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [118]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Multiple Formatting

Nothing prohibits using more than one conversion tool in the same print statement:

In [125]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


## Formatting with the `.format()` method

A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [126]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [127]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


#### 2. Inserted objects can be assigned keywords:

In [128]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [129]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with `.format()`

Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [140]:
print('{0:8} | {1:8}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:8}'.format('Apples', 3.))
print('{0:8} | {1:8}'.format('Oranges', 10))

Fruit    | Quantity
Apples   |      3.0
Oranges  |       10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [142]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [145]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [147]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

## Formatted String Literals (f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [148]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Pass `!r` to get the string representation:

In [149]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


#### Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [159]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [151]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [160]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

That is the basics of string formatting!

---

# 5. Lists

Earlier when discussing strings we introduced the concept of a *sequence* in Python. Lists can be thought of the most general version of a *sequence* in Python. Unlike strings, they are mutable, meaning the elements inside a list can be changed!

In this section we will learn about:
    
    1.) Creating lists
    2.) Indexing and Slicing Lists
    3.) Basic List Methods
    4.) Nesting Lists

Lists are constructed with brackets [] and commas separating every element in the list.

Let's go ahead and see how we can construct lists!

In [161]:
# Assign a list to an variable named my_list
my_list = [1,2,3]

We just created a list of integers, but lists can actually hold different object types. For example:

In [162]:
my_list = ['A string',23,100.232,'o']

Just like strings, the len() function will tell you how many items are in the sequence of the list.

In [163]:
len(my_list)

4

## Indexing and Slicing

Indexing and slicing work just like in strings. Let's make a new list to remind ourselves of how this works:

In [164]:
my_list = ['one','two','three',4,5]

In [165]:
# Grab element at index 0
my_list[0]

'one'

In [166]:
# Grab index 1 and everything past it
my_list[1:]

['two', 'three', 4, 5]

In [167]:
# Grab everything UP TO index 3
my_list[:3]

['one', 'two', 'three']

We can also use + to concatenate lists, just like we did for strings.

In [170]:
my_list + ['new item']

['one', 'two', 'three', 4, 5, 'new item']

Note: This doesn't actually change the original list!

In [171]:
my_list

['one', 'two', 'three', 4, 5]

You would have to reassign the list to make the change permanent.

In [172]:
# Reassign
my_list = my_list + ['add new item permanently']

In [175]:
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

We can also use the * for a duplication method similar to strings:

In [176]:
# Make the list double
my_list * 2

['one',
 'two',
 'three',
 4,
 5,
 'add new item permanently',
 'one',
 'two',
 'three',
 4,
 5,
 'add new item permanently']

In [177]:
# Again doubling not permanent
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

## Basic List Methods

If you are familiar with another programming language, you might start to draw parallels between arrays in another language and lists in Python. Lists in Python however, tend to be more flexible than arrays in other languages for a two good reasons: they have no fixed size (meaning we don't have to specify how big a list will be), and they have no fixed type constraint (like we've seen above).

Let's go ahead and explore some more special methods for lists:

In [180]:
# Create a new list
list1 = [1,2,3]

Use the **append** method to permanently add an item to the end of a list:

In [181]:
# Append
list1.append('append me!')

In [184]:
# Show
list1

[2, 3, 'append me!']

Use **pop** to "pop off" an item from the list. By default pop takes off the last index, but you can also specify which index to pop off. Let's see an example:

In [183]:
# Pop off the 0 indexed item
list1.pop(0)

1

In [185]:
# Show
list1

[2, 3, 'append me!']

In [186]:
# Assign the popped element, remember default popped index is -1
popped_item = list1.pop()

In [187]:
popped_item

'append me!'

In [188]:
# Show remaining list
list1

[2, 3]

It should also be noted that lists indexing will return an error if there is no element at that index. For example:

In [190]:
list1[100]

IndexError: list index out of range

We can use the **sort** method and the **reverse** methods to also effect your lists:

In [191]:
new_list = ['a','e','x','b','c']

In [192]:
#Show
new_list

['a', 'e', 'x', 'b', 'c']

In [193]:
# Use reverse to reverse order (this is permanent!)
new_list.reverse()

In [194]:
new_list

['c', 'b', 'x', 'e', 'a']

In [201]:
# Use sort to sort the list (in this case alphabetical order, but for numbers it will go ascending)
new_list.sort()

In [202]:
new_list

['a', 'b', 'c', 'e', 'x']

## Nesting Lists

A great feature of of Python data structures is that they support *nesting*. This means we can have data structures within data structures. For example: A list inside a list.

Let's see how this works!

In [203]:
# Let's make three lists
lst_1=[1,2,3]
lst_2=[4,5,6]
lst_3=[7,8,9]

# Make a list of lists to form a matrix
matrix = [lst_1,lst_2,lst_3]

In [204]:
# Show
matrix

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]

We can again use indexing to grab elements, but now there are two levels for the index. The items in the matrix object, and then the items inside that list!

In [205]:
# Grab first item in matrix object
matrix[0]

[1, 2, 3]

In [206]:
# Grab first item of the first item in the matrix object
matrix[0][0]

1

---

# 6. For Loops


A <code>for</code> loop acts as an iterator in Python; it goes through items that are in a *sequence* or any other iterable item. Objects that we've learned about that we can iterate over include strings, lists, tuples, and even built-in iterables for dictionaries, such as keys or values.

We've already seen the <code>for</code> statement a little bit in past lectures but now let's formalize our understanding.

Here's the general format for a <code>for</code> loop in Python:

    for item in object:
        statements to do stuff
    

The variable name used for the item is completely up to the coder, so use your best judgment for choosing a name that makes sense and you will be able to understand when revisiting your code. This item name can then be referenced inside your loop, for example if you wanted to use <code>if</code> statements to perform checks.

Let's go ahead and work through several example of <code>for</code> loops using a variety of data object types. We'll start simple and build more complexity later on.

## Example 1

Iterating through a list

In [207]:
# We'll learn how to automate this sort of list in the next lecture
list1 = [1,2,3,4,5,6,7,8,9,10]

In [208]:
for num in list1:
    print(num)

1
2
3
4
5
6
7
8
9
10


Great! Hopefully this makes sense. Now let's add an <code>if</code> statement to check for even numbers. We'll first introduce a new concept here -> **the modulo**.<br>
The modulo allows us to get the remainder in a division and uses the % symbol. For example:

In [209]:
17 % 5

2

This makes sense since 17 divided by 5 is 3 remainder 2. Let's see a few more quick examples:

In [210]:
# 3 Remainder 1
10 % 3

1

In [211]:
# 2 Remainder 4
18 % 7

4

In [212]:
# 2 no remainder
4 % 2

0

Notice that if a number is fully divisible with no remainder, the result of the modulo call is 0. We can use this to test for even numbers, since if a number modulo 2 is equal to 0, that means it is an even number!
Back to the <code>for</code> loops!


## Example 2

Let's print only the even numbers from that list!

In [213]:
for num in list1:
    if num % 2 == 0:
        print(num)

2
4
6
8
10


We could have also put an <code>else</code> statement in there:

In [214]:
for num in list1:
    if num % 2 == 0:
        print(num)
    else:
        print('Odd number')

Odd number
2
Odd number
4
Odd number
6
Odd number
8
Odd number
10


## Example 3

Another common idea during a <code>for</code> loop is keeping some sort of running tally during multiple loops. For example, let's create a <code>for</code> loop that sums up the list:

In [215]:
# Start sum at zero
list_sum = 0 

for num in list1:
    list_sum = list_sum + num

print(list_sum)

55


Great! Read over the above cell and make sure you understand fully what is going on. Also we could have implemented a <code>+=</code> to perform the addition towards the sum. For example:

In [216]:
# Start sum at zero
list_sum = 0 

for num in list1:
    list_sum += num

print(list_sum)

55


## Example 4

We've used <code>for</code> loops with lists, how about with strings? Remember strings are a sequence so when we iterate through them we will be accessing each item in that string.

In [218]:
for letter in 'This is a string.':
    print(letter)

T
h
i
s
 
i
s
 
a
 
s
t
r
i
n
g
.


## Example 5

Let's now look at how a <code>for</code> loop can be used with a tuple:

In [219]:
tup = (1,2,3,4,5)

for t in tup:
    print(t)

1
2
3
4
5


## Example 6

Tuples have a special quality when it comes to <code>for</code> loops. If you are iterating through a sequence that contains tuples, the item can actually be the tuple itself, this is an example of *tuple unpacking*. During the <code>for</code> loop we will be unpacking the tuple inside of a sequence and we can access the individual items inside that tuple!

In [220]:
list2 = [(2,4),(6,8),(10,12)]

In [222]:
for tup in list2:
    print(tup)

(2, 4)
(6, 8)
(10, 12)


In [225]:
# Now with unpacking!
for t1,t2 in list2:
    print(t1)

2
6
10


Cool! With tuples in a sequence we can access the items inside of them through unpacking! The reason this is important is because many objects will deliver their iterables through tuples. Let's start exploring iterating through Dictionaries to explore this further!

## Example 7

In [226]:
d = {'k1':1,'k2':2,'k3':3}

In [227]:
for item in d:
    print(item)

k1
k2
k3


Notice how this produces only the keys. So how can we get the values? Or both the keys and the values? 

We're going to introduce three new Dictionary methods: **.keys()**, **.values()** and **.items()**

In Python each of these methods return a *dictionary view object*. It supports operations like membership test and iteration, but its contents are not independent of the original dictionary – it is only a view. Let's see it in action:

In [ ]:
# Create a dictionary view object
d.items()

dict_items([('k1', 1), ('k2', 2), ('k3', 3)])

Since the .items() method supports iteration, we can perform *dictionary unpacking* to separate keys and values just as we did in the previous examples.

In [228]:
# Dictionary unpacking
for k,v in d.items():
    print(k)
    print(v) 

k1
1
k2
2
k3
3


If you want to obtain a true list of keys, values, or key/value tuples, you can *cast* the view as a list:

In [ ]:
list(d.keys())

['k1', 'k2', 'k3']

Remember that dictionaries are unordered, and that keys and values come back in arbitrary order. You can obtain a sorted list using sorted():

In [ ]:
sorted(d.values())

[1, 2, 3]

## Conclusion

We've learned how to use for loops to iterate through tuples, lists, strings, and dictionaries. It will be an important tool for us, so make sure you know it well and understood the above examples.

[More resources](http://www.tutorialspoint.com/python/python_for_loop.htm)